SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestRPCTimeout/SpatialTestRPCTimeout.cpp

```diff
@@ -0,0 +1,89 @@
+// Copyright (c) Improbable Worlds Ltd, All Rights Reserved
+
+#include "SpatialTestRPCTimeout.h"
+#include "Components/BoxComponent.h"
+#include "GameFramework/Character.h"
+#include "GameFramework/PlayerController.h"
+#include "SpatialFunctionalTestFlowController.h"
+#include "SpatialTestRPCTimeoutGameMode.h"
+#include "SpatialTestRPCTimeoutPlayerController.h"
+
+/**
+ * This test ensures that RPC calls with unresolved parameters are queued until their parameters are correctly resolved.
+ * This test contains 1 Server and 2 Client workers running in multiple processes.
+ *
+ * The flow is as follows:
+ * - Setup:
+ *  - Launch at least one client in a separate process
+ * - Test:
+ *  - All clients launched outside of the editor process must ensure that the referenced material asset is not in memory at the start of the
+ * test.
+ *  - All clients launched outside of the editor process must successfully resolve the material asset before passing it into a client RPC
+ * function.
+ */
+
+ASpatialTestRPCTimeout::ASpatialTestRPCTimeout()
+	: Super()
+{
+	Author = "Iwan";
+	Description = TEXT(
+		"This test calls an RPC with an asset that was softly referenced and check that it will be asynchronously loaded with a timeout of "
+		"0.");
+}
+
+void ASpatialTestRPCTimeout::CreateCustomContentForMap()
+{
+	GetWorldSettingsForMap()->DefaultGameMode = ASpatialTestRPCTimeoutGameMode::StaticClass();
+}
+
+void ASpatialTestRPCTimeout::PrepareTest()
+{
+	Super::PrepareTest();
+
+	AddStep(
+		TEXT("Check that the material was not initially loaded on non-editor clients"), FWorkerDefinition::AllClients, nullptr, nullptr,
+		[this](float DeltaTime) {
+			if (!GIsEditor)
+			{
+				ACharacter* TestCharacter = Cast<ACharacter>(GetLocalFlowPawn());
+				ASpatialTestRPCTimeoutPlayerController* TestController =
+					Cast<ASpatialTestRPCTimeoutPlayerController>(TestCharacter->GetController());
+
+				if (TestController && TestCharacter)
+				{
+					UMaterial* Material = TestController->SoftMaterialPtr.Get();
+
+					RequireTrue(Material == nullptr, TEXT("The soft-pointed material must not be in memory at the start of the test."));
+					UE_LOG(LogTemp, Display, TEXT("A"));
+
+					FinishStep();
+				}
+			}
+			else
+			{
+				FinishStep();
+			}
+		},
+		2.0f);
+
+	AddStep(
+		TEXT("Check that the material is correctly loaded after about 2 seconds delay. We should wait "), FWorkerDefinition::AllClients,
+		nullptr, nullptr,
+		[this](float DeltaTime) {
+			if (!GIsEditor)
+			{
+				ACharacter* TestCharacter = Cast<ACharacter>(GetLocalFlowPawn());
+				ASpatialTestRPCTimeoutPlayerController* TestController =
+					Cast<ASpatialTestRPCTimeoutPlayerController>(TestCharacter->GetController());
+
+				RequireTrue(TestController->IsSuccessfullyResolved(),
+							TEXT("The soft-pointed material is synchronously loaded into the non-editor process."));
+				FinishStep();
+			}
+			else
+			{
+				FinishStep();
+			}
+		},
+		5.0f);
```

Comment:
Any chance we can increase the chance this succeeds under load  by having some exponential backoff? Or something

Reply:
The 2 seconds comment is added not because we expect any particular stress during this test, but because the player controller is programmed to explicitly wait 2 seconds before "starting" the test (this was the case in the original RPC Timeout TestGym to make sure to characters are correctly possessed before we assign materials to them)

We can make this safer by increasing the timeout length of this step, or, if we think a backoff would be safer, I can program this in the logic `RPCTimeoutPlayerController`?

Reply:
Ah did you get round to this or?

Reply:
The former yes, and I think Miron's suggestion for the other step helped quite a bit with this as well!

Now the timing of this step should line up better with the delay because the previous step requires us to wait a few seconds. Ran this test repeatedly in combination with others and it seems to be passing consistently.